mlflow/telemetry/client.py

```diff
@@ -1,33 +1,251 @@
+import atexit
+import json
 import logging
+import threading
+import time
+from concurrent.futures import FIRST_COMPLETED, ThreadPoolExecutor, wait
+from dataclasses import asdict
+from queue import Empty, Full, Queue
+from typing import Optional
 
-from mlflow.telemetry.schemas import Record, TelemetryInfo
+import requests
+
+from mlflow.environment_variables import (
+    _MLFLOW_TELEMETRY_BATCH_SIZE,
+    _MLFLOW_TELEMETRY_BATCH_TIME_INTERVAL,
+    _MLFLOW_TELEMETRY_MAX_QUEUE_SIZE,
+    _MLFLOW_TELEMETRY_MAX_WORKERS,
+)
+from mlflow.telemetry.schemas import APIRecord, TelemetryInfo, TelemetryRecord
 from mlflow.telemetry.utils import is_telemetry_disabled
-from mlflow.tracking._tracking_service.utils import _get_store
 
 _logger = logging.getLogger(__name__)
+# TODO: update this url with a custom domain
+TELEMETRY_URL = "https://twqhrx9tai.execute-api.us-west-2.amazonaws.com/test/telemetry"
 
 
 class TelemetryClient:
     def __init__(self):
         self.info = TelemetryInfo()
-        self.records: list[Record] = []
+        self.telemetry_url = TELEMETRY_URL
+        self._queue: Queue[list[APIRecord]] = Queue(maxsize=_MLFLOW_TELEMETRY_MAX_QUEUE_SIZE.get())
+        self._lock = threading.RLock()
+        self._max_workers = _MLFLOW_TELEMETRY_MAX_WORKERS.get()
+
+        # Thread event that indicates the queue should stop processing tasks
+        self._stop_event = threading.Event()
+        self._is_active = False
+        self._atexit_callback_registered = False
+
+        self._active_tasks = set()
 
-    # temp helper function to test
-    # TODO: update this in next PR
-    def add_record(self, record: Record):
+        self._batch_size = _MLFLOW_TELEMETRY_BATCH_SIZE.get()
+        self._batch_time_interval = _MLFLOW_TELEMETRY_BATCH_TIME_INTERVAL.get()
+        self._pending_records: list[TelemetryRecord] = []
+        self._last_batch_time = time.time()
+        self._batch_lock = threading.Lock()
+
+    def add_record(self, record: APIRecord):
         """
-        Add a record to the queue to be sent to the telemetry server.
+        Add a record to be batched and sent to the telemetry server.
         """
-        self.records.append(record)
+        if not self.is_active():
+            self.activate()
+
+        # If stop event is set, don't add new records
+        if self._stop_event.is_set():
+            _logger.debug("Telemetry is stopped, skipping adding record")
+            return
+
+        with self._batch_lock:
+            data = self._generate_telemetry_record(record)
```

Comment:
nit: I think record generation can happen outside batch lock to minimize the scope of lock